src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets

```
@@ -339,23 +341,37 @@ The .NET Foundation licenses this file to you under the MIT license.
-      <CompileEmccArgs Condition="'$(NativeDebugSymbols)' == 'true'">$(CompileEmccArgs) -g3</CompileEmccArgs>
-      <!-- TODO-LLVM: remove when we update to Emscripten with https://github.com/emscripten-core/emscripten/pull/18443 -->
-      <CompileEmccArgs>$(CompileEmccArgs) -s USE_SDL=0 $(EmccOptimizationSetting)</CompileEmccArgs>
+    <PropertyGroup Condition="'$(_targetOS)' == 'browser'">
+      <CompileWasmArgs>-c</CompileWasmArgs>
+      <CompileWasmArgs>$(CompileWasmArgs) -s DISABLE_EXCEPTION_CATCHING=0</CompileWasmArgs>
+      <CompileWasmArgs Condition="'$(NativeDebugSymbols)' == 'true'">$(CompileWasmArgs) -g3</CompileWasmArgs>
+      <!-- TODO-LLVM: remove -s USE_SDL=0 when we update to Emscripten with https://github.com/emscripten-core/emscripten/pull/18443 -->
+      <CompileWasmArgs>$(CompileWasmArgs) -s USE_SDL=0 $(EmccOptimizationSetting)</CompileWasmArgs>
 
       <ScriptExt Condition="'$(OS)' == 'Windows_NT'">.bat</ScriptExt>
-      <EmccPath>&quot;$(EMSDK)/upstream/emscripten/emcc$(ScriptExt)&quot;</EmccPath>
+      <WasmCompilerPath>&quot;$(EMSDK)/upstream/emscripten/emcc$(ScriptExt)&quot;</WasmCompilerPath>
+      <WasmLinkerPath>$(WasmCompilerPath)</WasmLinkerPath>
     </PropertyGroup>
 
-    <Exec Command="$(EmccPath) &quot;%(LlvmObjects.Identity)&quot; -o &quot;%(LlvmObjects.NativeObject)&quot; $(CompileEmccArgs)" Condition="'$(EMSDK)' != ''" />
-    <Message Text="Emscripten not found, not linking WebAssembly. To enable WebAssembly linking, install Emscripten and ensure the EMSDK environment variable points to the directory containing upstream/emscripten/emcc.bat"
-             Condition="'$(EMSDK)' == ''" />
+   <PropertyGroup Condition="'$(_targetOS)' == 'wasi'">
+      <CompileWasmArgs>-fvisibility=default -mllvm -combiner-global-alias-analysis=false -mllvm -disable-lsr --sysroot=&quot;$(WASI_SDK_PATH)/share/wasi-sysroot&quot; -pthread -target wasm32-wasi-threads -c</CompileWasmArgs>
+      <CompileWasmArgs Condition="'$(NativeDebugSymbols)' != 'true'">$(CompileWasmArgs) -O2</CompileWasmArgs>
+      <CompileWasmArgs Condition="'$(NativeDebugSymbols)' == 'true'">$(CompileWasmArgs) -g3</CompileWasmArgs>
+
+      <ExeExt Condition="'$(OS)' == 'Windows_NT'">.exe</ExeExt>
+      <!-- using Emscripten's clang++ because of a crash in wasi-sdk's clang++ (https://github.com/WebAssembly/wasi-sdk/issues/326) -->
+      <WasmCompilerPath>&quot;$(EMSDK)/upstream/bin/clang++$(ExeExt)&quot;</WasmCompilerPath>
+      <WasmLinkerPath>&quot;$(WASI_SDK_PATH)/bin/clang&quot;</WasmLinkerPath>
+    </PropertyGroup>
+
+    <Exec Command="$(WasmCompilerPath) &quot;%(LlvmObjects.Identity)&quot; -o &quot;%(LlvmObjects.NativeObject)&quot; $(CompileWasmArgs)" Condition="'$(EMSDK)' != ''" />
+    <Message Text="Emscripten not found, not linking WebAssembly for Wasi. To enable WebAssembly linking, install Emscripten and ensure the EMSDK environment variable points to the directory containing upstream/emscripten/emcc.bat"
```

```suggestion
    <Message Text="Emscripten not found, not linking WebAssembly. To enable WebAssembly linking, install Emscripten and ensure the EMSDK environment variable points to the directory containing upstream/emscripten/emcc.bat"
```
This is under `'$(_targetOS)' == 'browser'`.